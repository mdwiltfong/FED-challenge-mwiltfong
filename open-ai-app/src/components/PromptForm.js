import React, { useState } from "react";
import axios from "axios";
import { v4 as uuid } from "uuid";
import Response from "./Response";
const PromptForm = () => {
  const [formData, setFormData] = useState();
  const [responses, setResponses] = useState([]);
  const handleSubmit = async (e) => {
    try {
      e.preventDefault();
      const data = {
        id: uuid(),
        prompt: formData,
      };
      const {
        data: { prompt, response },
      } = await axios.post("http://localhost:8001/create_completion", data);

      setResponses((prevResponses) => [
        ...prevResponses,
        { id: uuid(), response: response, prompt: prompt },
      ]);
    } catch (error) {
      console.log(error);
    }
  };
  const handleChange = (e) => {
    const { value } = e.target;
    setFormData(value);
  };
  let responseContent;
  if (responses.length > 0) {
    responseContent = (
      <>
        <h2>Responses</h2>
        <ul>
          {responses.map((response) => (
            <Response
              key={response.id}
              prompt={response.prompt}
              response={response.response}
            />
          ))}
        </ul>
      </>
    );
  } else {
    responseContent = (
      <>
        <h2>Responses</h2>
        <p>Submit a prompt to get a response from the Open AI API!</p>
      </>
    );
  }

  return (
    <>
      <h1>Fun with AI</h1>
      <p>Enter prompt</p>
      <form
        style={{
          maxWidth: "80%",
        }}
        className="d-flex flex-column"
        data-testid="prompt-form"
        onSubmit={handleSubmit}
      >
        <textarea onChange={handleChange} rows="10" value={formData}></textarea>
        <button
          style={{
            margin: "0 auto",
            marginRight: "0px",
            maxWidth: "80px",
            backgroundColor: "#0043C8",
          }}
          className="btn btn-primary"
        >
          Submit
        </button>
      </form>
      {responseContent}
    </>
  );
};

export default PromptForm;
