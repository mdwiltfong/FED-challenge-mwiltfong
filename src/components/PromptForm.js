import React, { useState, useEffect } from "react";
import axios from "axios";
import { v4 as uuid } from "uuid";
import Response from "./Response";
import retrieveStorage from "../helper_functions/retrieveStorage";
import populateStorage from "../helper_functions/populateStorage";
import "./styles/PromptForm.css";

const PromptForm = () => {
  const INITIAL_STATE = {
    engine: "text-curie-001",
    textarea: "",
  };
  const [formData, setFormData] = useState(INITIAL_STATE);
  const [responses, setResponses] = useState([]);
  useEffect(() => {
    const storedResponses = retrieveStorage() || null;
    if (storedResponses) {
      setResponses(storedResponses);
    }
  }, []);
  const handleSubmit = async (e) => {
    try {
      e.preventDefault();
      const data = {
        id: uuid(),
        prompt: formData.textarea,
      };
      const config = {
        params: { engine: formData.engine },
      };

      const {
        data: { prompt, response },
      } = await axios.post(
        "https://openai-app-mw.herokuapp.com/create_completion",
        data,
        config
      );
      populateStorage({ id: uuid(), prompt: prompt, response: response });
      setResponses((prevResponses) => [
        { id: uuid(), response: response, prompt: prompt },
        ...prevResponses,
      ]);
    } catch (error) {
      console.log(error);
    }
  };
  const handleChangeTextArea = (e) => {
    const { name, value } = e.target;

    setFormData((prevState) => ({
      ...prevState,
      [name]: value,
    }));
  };
  const handleChangeInput = (e) => {
    const { name, value } = e.target;
    setFormData((prevState) => ({
      ...prevState,
      [name]: value,
    }));
    console.log(formData);
  };
  let responseContent;
  if (responses.length > 0) {
    responseContent = (
      <>
        <h2>Responses</h2>
        <ul style={{ listStyle: "none", padding: "0" }}>
          {responses.map((response) => (
            <Response
              key={response.id}
              prompt={response.prompt}
              response={response.response}
            />
          ))}
        </ul>
      </>
    );
  } else {
    responseContent = (
      <>
        <h2 aria-label="Responses">Responses</h2>
        <p aria-label="Submit a prompt to get a response from the Open AI API!">
          Submit a prompt to get a response from the Open AI API!
        </p>
      </>
    );
  }

  return (
    <>
      <h1 aria-label="Fun With AI">Fun with AI</h1>
      <a id="skip-nav" className="screenreader-text" href="#main-content">
        Skip Navigation or Skip to Content
      </a>
      <form
        style={{
          maxWidth: "100%",
        }}
        className="d-flex flex-column"
        data-testid="prompt-form"
        onSubmit={handleSubmit}
        role="form"
        aria-label="Text Area Form"
      >
        <label htmlFor="engine" aria-label="Select an AI Engine">
          Select an AI Engine{" "}
        </label>
        <select
          defaultValue={formData.engine}
          onChange={handleChangeInput}
          name="engine"
          style={{
            width: "200px",
          }}
        >
          <option value="text-davinci-002">text-davinci-002</option>
          <option value="text-curie-001">text-curie-001</option>
          <option value="text-babbage-001">text-babbage-001</option>
          <option value="text-ada-001">text-ada-001</option>
        </select>
        <label htmlFor="textarea">Enter your Prompt</label>
        <textarea
          onChange={handleChangeTextArea}
          rows="10"
          value={formData.textarea}
          name="textarea"
          id="main-content"
          role="main"
          aria-label="Text Area"
        ></textarea>
        <button
          style={{
            margin: "0 auto",
            marginRight: "0px",
            maxWidth: "80px",
            backgroundColor: "#0043C8",
          }}
          className="btn btn-primary"
          role="button"
          aria-label="Submit Prompt Button to OpenAI API"
        >
          Submit
        </button>
      </form>
      {responseContent}
    </>
  );
};

export default PromptForm;
