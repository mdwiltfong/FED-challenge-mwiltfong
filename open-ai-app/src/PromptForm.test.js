import PromptForm from "./components/PromptForm";
import { render } from "@testing-library/react";
describe("PromptForm component Tests", () => {
  it("Smoke Tests", () => {
    render(<PromptForm />);
  });
  it("Snapshot Tests", () => {
    const screen = render(<PromptForm />);
    expect(screen.asFragment()).toMatchSnapshot();
  });
  it("Renders a form with a textarea, and submit button", () => {
    const screen = render(<PromptForm />);

    const promptForm = screen.container.getElementsByTagName("form");
    const textArea = screen.container.getElementsByTagName("textarea");
    const submitBtn = screen.getByText("Submit");
    expect(promptForm).toBeInTheDocument();
    expect(textArea).toBeInTheDocument();
    expect(submitBtn).toBeInTheDocument();
  });
});
