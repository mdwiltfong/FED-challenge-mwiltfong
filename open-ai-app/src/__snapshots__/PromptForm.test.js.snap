// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptForm component Tests Snapshot Tests 1`] = `
<DocumentFragment>
  <h1>
    Fun with AI
  </h1>
  <p>
    Enter prompt
  </p>
  <form
    class="d-flex flex-column"
    data-testid="prompt-form"
    style="max-width: 80%;"
  >
    <textarea
      rows="10"
    />
    <button
      class="btn btn-primary"
      style="margin: 0px 0px 0px auto; max-width: 80px; background-color: rgb(0, 67, 200);"
    >
      Submit
    </button>
  </form>
  <h2>
    Responses
  </h2>
  <p>
    Submit a prompt to get a response from the Open AI API!
  </p>
</DocumentFragment>
`;
